File: bindings/swift/Sources/ZeroDevAA/Context.swift
import CZeroDevAA

public final class Context: @unchecked Sendable {
    let ptr: OpaquePointer

    public init(projectID: String, rpcURL: String = "", bundlerURL: String = "", chainID: UInt64, gasMiddleware: GasMiddleware, paymasterMiddleware: PaymasterMiddleware = .zeroDev) throws {
        var out: OpaquePointer?
        let status = aa_context_create(projectID, rpcURL, bundlerURL, chainID, &out)
        try checkResult(status)
        guard let p = out else { throw AAError.nullOutPtr }
        self.ptr = p

        switch gasMiddleware {
        case .zeroDev:
            try checkResult(aa_context_set_gas_middleware(ptr, aa_gas_zerodev))
        }

        switch paymasterMiddleware {
        case .zeroDev:
            try checkResult(aa_context_set_paymaster_middleware(ptr, aa_paymaster_zerodev))
        case .none:
            break
        }

        // Use URLSession for all HTTP on Apple platforms (Zig's TLS fails on iOS)
        useURLSessionTransport()
    }

    /// Create a Kernel smart account.
    ///
    /// When `address` is `nil` (the default), the sender is derived
    /// counterfactually via CREATE2 from `(signer, version, index)`. When
    /// supplied, the account's sender is pinned to that address
    /// (migration path for kernel-version upgrades, or operating a
    /// pre-existing wallet whose CREATE2 salt this SDK no longer
    /// computes). Pinning affects the sender only; factory init_code is
    /// still emitted on the first UserOp exactly as it would be for a
    /// counterfactually-derived account (governed by the EntryPoint
    /// nonce). Callers pinning an already-deployed account with
    /// EntryPoint nonce 0 (rare — funded but never used) should drop the
    /// factory bytes via the low-level UserOp API.
    public func newAccount(
        signer: Signer,
        version: KernelVersion,
        index: UInt32 = 0,
        address: [UInt8]? = nil
    ) throws -> Account {
        if let a = address {
            precondition(a.count == 20, "address must be 20 bytes")
        }
        var out: OpaquePointer?
        let status = address?.withUnsafeBufferPointer { buf in
            aa_account_create(
                ptr,
                signer.ptr,
                aa_kernel_version(rawValue: UInt32(version.rawValue)),
                index,
                buf.baseAddress,
                &out
            )
        } ?? aa_account_create(
            ptr,
            signer.ptr,
            aa_kernel_version(rawValue: UInt32(version.rawValue)),
            index,
            nil,
            &out
        )
        try checkResult(status)
        guard let p = out else { throw AAError.nullOutPtr }
        return Account(ptr: p, context: self, signer: signer)
    }

    /// Create a Kernel smart account using EIP-7702 delegation.
    ///
    /// The account's address is the signer's EOA address — there is no
    /// CREATE2, no init code, and no index parameter. On the first
    /// UserOperation the SDK signs an authorization tuple
    /// `(chainId, Kernel implementation for version, EOA nonce)` and attaches
    /// it via the `eip7702Auth` field; subsequent UserOps skip the
    /// authorization once delegation is installed on-chain.
    ///
    /// Today only `.v3_3` supports EIP-7702.
    public func newAccount7702(signer: Signer, version: KernelVersion) throws -> Account {
        var out: OpaquePointer?
        let status = aa_context_new_account_7702(ptr, signer.ptr, aa_kernel_version(rawValue: UInt32(version.rawValue)), &out)
        try checkResult(status)
        guard let p = out else { throw AAError.nullOutPtr }
        return Account(ptr: p, context: self, signer: signer)
    }

    deinit {
        aa_context_destroy(ptr)
    }
}

File: bindings/kotlin/src/main/kotlin/dev/zerodev/aa/Context.kt
package dev.zerodev.aa

class Context private constructor(internal val ptr: Long) : AutoCloseable {
    private var closed = false

    companion object {
        fun create(
            projectId: String,
            rpcUrl: String = "",
            bundlerUrl: String = "",
            chainId: Long = 11155111,
            gasMiddleware: GasMiddleware = GasMiddleware.ZERODEV,
            paymasterMiddleware: PaymasterMiddleware = PaymasterMiddleware.ZERODEV,
        ): Context {
            val out = LongArray(1)
            checkStatus(NativeLib.nContextCreate(projectId, rpcUrl, bundlerUrl, chainId, out))
            val ctx = out[0]

            when (gasMiddleware) {
                GasMiddleware.ZERODEV -> checkStatus(NativeLib.nContextSetGasZeroDev(ctx))
                GasMiddleware.PIMLICO -> checkStatus(NativeLib.nContextSetGasPimlico(ctx))
            }

            when (paymasterMiddleware) {
                PaymasterMiddleware.ZERODEV -> checkStatus(NativeLib.nContextSetPaymasterZeroDev(ctx))
                PaymasterMiddleware.NONE -> { /* No paymaster — send unsponsored */ }
            }

            return Context(ctx)
        }
    }

    /**
     * Create a Kernel smart account.
     *
     * When [address] is null (the default), the sender is derived
     * counterfactually via CREATE2 from `(signer, version, index)`. When
     * supplied as a 20-byte array, the account's sender is pinned to that
     * address (migration path for kernel-version upgrades or legacy
     * wallets whose CREATE2 salt this SDK no longer computes). Pinning
     * affects the sender only; factory init_code is still emitted on the
     * first UserOp exactly as it would be for a counterfactually-derived
     * account (governed by the EntryPoint nonce). Callers pinning an
     * already-deployed account with EntryPoint nonce 0 (rare — funded but
     * never used) should drop the factory bytes via the low-level UserOp
     * API.
     */
    fun newAccount(
        signer: Signer,
        version: KernelVersion,
        index: Int = 0,
        address: ByteArray? = null,
    ): Account {
        check(!closed) { "Context is closed" }
        if (address != null) {
            require(address.size == 20) { "address must be 20 bytes, got ${address.size}" }
        }
        val out = LongArray(1)
        checkStatus(NativeLib.nAccountCreate(ptr, signer.ptr, version.code, index, address, out))
        return Account(out[0], this, signer)
    }

    /**
     * Create an EIP-7702 account. The account address IS the signer's EOA —
     * no CREATE2, no init code, no index. On the first UserOp the SDK signs
     * an authorization tuple and attaches it via `eip7702Auth`; subsequent
     * ops reuse the delegation. Today only [KernelVersion.V3_3] supports 7702.
     */
    fun newAccount7702(
        signer: Signer,
        version: KernelVersion,
    ): Account {
        check(!closed) { "Context is closed" }
        val out = LongArray(1)
        checkStatus(NativeLib.nAccountCreate7702(ptr, signer.ptr, version.code, out))
        return Account(out[0], this, signer)
    }

    override fun close() {
        if (!closed) {
            NativeLib.nContextDestroy(ptr)
            closed = true
        }
    }
}
